README.md
# CPUFreq - Swift 3 and 4 bindings to CPUFreq in Linux sysfs

The project provides the `CPUFreq` Swift package - a thin wrapper around the `CPUFreq` Linux interface in sysfs for getting and setting DVFS-related configurations.
Its purpose is to abstract programs from file I/O and data parsing/conversion.
It does not provide core/domain/socket discovery or DVFS management logic.

For Linux kernel documentation on CPUFreq, see: https://www.kernel.org/doc/Documentation/cpu-freq/user-guide.txt


## Building

To build, run:

```sh
swift build
```


## Usage

A simple example of setting DVFS frequencies with the `userspace` governor:

```Swift
import CPUFreq

// in practice, core IDs could be discovered dynamically using other means...
let bindings: [CPUFreq] = [ CPUFreq(cpu: 0), CPUFreq(cpu: 1) ]

// populate "availableFreqs" array
let availableFreqs: [UInt32] = bindings[0].getScalingAvailableFrequencies()!
// won't use that file again, so we can manually close it to free up resources
bindings[0].closeFile(file: CPUFreqFile.scalingAvailableFrequencies)

// set "userspace" governor on all cores
for cpu in bindings {
    let _ = cpu.setScalingGovernor(governor: "userspace")
    // again, close the file manually to free up resources
    cpu.closeFile(file: CPUFreqFile.scalingGovernor)
}

// do application work, breaking from loop when finished...
while (true) {
    // in practice, a new frequency would be determined dynamically by some scheduling logic...
    let freq = availableFreqs[0];
    for cpu in bindings {
        let _ = cpu.setScalingSetspeed(freq: freq);
    }
}

// once bindings go out of scope, any open files (e.g. CPUFreqFile.scalingSetspeed) are closed automatically
```

The files `scaling_governor`, `scaling_min_freq`, and `scaling_max_freq` are opened as read/write by default, but may be opened as read-only by specifying `readOnly: True` as a function parameter when they are first accessed.
The file `scaling_setspeed` is opened as write-only.
All other files are opened as read-only.
Users are responsible for ensuring that their programs have correct read/write privileges.


## Project Source

Find this and related project sources at the [powercap organization on GitHub](https://github.com/powercap).  
This project originates at: https://github.com/powercap/CPUFreq-Swift

Bug reports and pull requests are welcome.


RELEASES.md
# Release Notes

## [Unreleased]
### Added
* AUTHORS, LICENSE, and RELEASES.md files

## v1.0.0 - 2017-06-22
### Added
* Initial public release


[Unreleased]: https://github.com/powercap/CPUFreq-Swift/compare/v1.0.0...HEAD


Tests/LinuxMain.swift
import XCTest
@testable import CPUFreqTests

XCTMain([
    testCase(CPUFreqTests.allTests),
])


Tests/CPUFreqTests/CPUFreqTests.swift
import XCTest
@testable import CPUFreq

class CPUFreqTests: XCTestCase {

    func testCPUFreqFile() {
        // all files accounted for
        XCTAssert(CPUFreqFile.count == 15)
        // file names
        XCTAssert(CPUFreqFile.affectedCpus.getFileName() == "affected_cpus")
        XCTAssert(CPUFreqFile.biosLimit.getFileName() == "bios_limit")
        XCTAssert(CPUFreqFile.cpuinfoCurFreq.getFileName() == "cpuinfo_cur_freq")
        XCTAssert(CPUFreqFile.cpuinfoMaxFreq.getFileName() == "cpuinfo_max_freq")
        XCTAssert(CPUFreqFile.cpuinfoMinFreq.getFileName() == "cpuinfo_min_freq")
        XCTAssert(CPUFreqFile.cpuinfoTransitionLatency.getFileName() == "cpuinfo_transition_latency")
        XCTAssert(CPUFreqFile.relatedCpus.getFileName() == "related_cpus")
        XCTAssert(CPUFreqFile.scalingAvailableFrequencies.getFileName() == "scaling_available_frequencies")
        XCTAssert(CPUFreqFile.scalingAvailableGovernors.getFileName() == "scaling_available_governors")
        XCTAssert(CPUFreqFile.scalingCurFreq.getFileName() == "scaling_cur_freq")
        XCTAssert(CPUFreqFile.scalingDriver.getFileName() == "scaling_driver")
        XCTAssert(CPUFreqFile.scalingGovernor.getFileName() == "scaling_governor")
        XCTAssert(CPUFreqFile.scalingMaxFreq.getFileName() == "scaling_max_freq")
        XCTAssert(CPUFreqFile.scalingMinFreq.getFileName() == "scaling_min_freq")
        XCTAssert(CPUFreqFile.scalingSetspeed.getFileName() == "scaling_setspeed")
        // from file names
        XCTAssert(CPUFreqFile(fileName: "affected_cpus") == CPUFreqFile.affectedCpus)
        XCTAssert(CPUFreqFile(fileName: "bios_limit") == CPUFreqFile.biosLimit)
        XCTAssert(CPUFreqFile(fileName: "cpuinfo_cur_freq") == CPUFreqFile.cpuinfoCurFreq)
        XCTAssert(CPUFreqFile(fileName: "cpuinfo_max_freq") == CPUFreqFile.cpuinfoMaxFreq)
        XCTAssert(CPUFreqFile(fileName: "cpuinfo_min_freq") == CPUFreqFile.cpuinfoMinFreq)
        XCTAssert(CPUFreqFile(fileName: "cpuinfo_transition_latency") == CPUFreqFile.cpuinfoTransitionLatency)
        XCTAssert(CPUFreqFile(fileName: "related_cpus") == CPUFreqFile.relatedCpus)
        XCTAssert(CPUFreqFile(fileName: "scaling_available_frequencies") == CPUFreqFile.scalingAvailableFrequencies)
        XCTAssert(CPUFreqFile(fileName: "scaling_available_governors") == CPUFreqFile.scalingAvailableGovernors)
        XCTAssert(CPUFreqFile(fileName: "scaling_cur_freq") == CPUFreqFile.scalingCurFreq)
        XCTAssert(CPUFreqFile(fileName: "scaling_driver") == CPUFreqFile.scalingDriver)
        XCTAssert(CPUFreqFile(fileName: "scaling_governor") == CPUFreqFile.scalingGovernor)
        XCTAssert(CPUFreqFile(fileName: "scaling_max_freq") == CPUFreqFile.scalingMaxFreq)
        XCTAssert(CPUFreqFile(fileName: "scaling_min_freq") == CPUFreqFile.scalingMinFreq)
        XCTAssert(CPUFreqFile(fileName: "scaling_setspeed") == CPUFreqFile.scalingSetspeed)
        // path
        XCTAssert(CPUFreqFile.affectedCpus.getPath(cpu: 0) == "/sys/devices/system/cpu/cpu0/cpufreq/affected_cpus")
        XCTAssert(CPUFreqFile.affectedCpus.getPath(cpu: 1) == "/sys/devices/system/cpu/cpu1/cpufreq/affected_cpus")
    }

    func testCPUFreq() {
        let c = CPUFreq(cpu: 0)
        print(String(describing: c.getAffectedCpus()))
        print(String(describing: c.getBiosLimit()))
        print(String(describing: c.getCpuinfoCurFreq()))
        print(String(describing: c.getCpuinfoMaxFreq()))
        print(String(describing: c.getCpuinfoMinFreq()))
        print(String(describing: c.getCpuinfoTransitionLatency()))
        print(String(describing: c.getRelatedCpus()))
        print(String(describing: c.getScalingAvailableFrequencies()))
        print(String(describing: c.getScalingAvailableGovernors()))
        print(String(describing: c.getScalingCurFreq()))
        print(String(describing: c.getScalingDriver()))
        print(String(describing: c.getScalingGovernor(readOnly: true)))
        print(String(describing: c.getScalingMaxFreq(readOnly: true)))
        print(String(describing: c.getScalingMinFreq(readOnly: true)))
    }


    static var allTests = [
        ("testCPUFreqFile", testCPUFreqFile),
        ("testCPUFreq", testCPUFreq),
    ]
}


Sources/CPUFreq.swift
/**
 * Bindings to cpufreq in Linux sysfs.
 * Files that are read/write can be opened as read only so unprivileged applications can still read values.
 *
 * @author Connor Imes
 * @date 2017-06-15
 */

import Foundation

fileprivate extension FileHandle {

    func readString() -> String? {
        seek(toFileOffset: 0)
        // must replace newline characters for conversions to integer types to work
        return String(data: readDataToEndOfFile(), encoding: String.Encoding.ascii)?.replacingOccurrences(of: "\n", with: "")
    }

    func readStringArr() -> [String]? {
        return readString()?.components(separatedBy: " ")
    }

    func readU32() -> UInt32? {
        if let s: String = readString() {
            return UInt32(s)
        }
        return nil
    }

    func readU32Arr() -> [UInt32]? {
        return readStringArr()?.flatMap{UInt32($0)}
    }

    func writeString(_ s: String) -> Bool {
        if let data = s.data(using: String.Encoding.ascii, allowLossyConversion: false) {
            seek(toFileOffset: 0)
            write(data)
            return true
        }
        return false
    }

    func writeU32(_ u: UInt32) -> Bool {
        return writeString("\(u)")
    }
}

/// Bindings for CPUFreq that manage file I/O and data parsing/conversion.
/**
 Manages reading and writing files for files in the CPUFreq sysfs interface in Linux.
 The class automatically opens files when a read or write is requested.
 Files are automatically closed when the class instance is deallocated.
 A function is also available for users to close individual files without deallocating the class instance.
*/
public class CPUFreq {
    /// The CPU that these bindings operate on
    public let cpu: UInt32
    /// FileHandle instances for each potential file
    private var files: [FileHandle?] = [FileHandle?](repeating: nil, count: CPUFreqFile.count)

    /// Instantiate for a CPU - cpu0 is used by default.
    public init(cpu: UInt32 = 0) {
        self.cpu = cpu
    }

    /// Close a file if it is open.
    public func closeFile(file: CPUFreqFile) {
        files[file.rawValue] = nil
    }

    private func openForR(_ f: CPUFreqFile) {
        if files[f.rawValue] == nil {
            files[f.rawValue] = FileHandle(forReadingAtPath: f.getPath(cpu: cpu))
        }
    }

    private func openForW(_ f: CPUFreqFile) {
        if files[f.rawValue] == nil {
            files[f.rawValue] = FileHandle(forWritingAtPath: f.getPath(cpu: cpu))
        }
    }

    private func openForRW(_ f: CPUFreqFile, _ readOnly: Bool) {
        if readOnly {
            openForR(f)
        } else if files[f.rawValue] == nil {
            files[f.rawValue] = FileHandle(forUpdatingAtPath: f.getPath(cpu: cpu))
        }
    }

    /// Get the affected CPUs.
    public func getAffectedCpus() -> [UInt32]? {
        openForR(CPUFreqFile.affectedCpus)
        return files[CPUFreqFile.affectedCpus.rawValue]?.readU32Arr()
    }

    /// Get the BIOS limit.
    public func getBiosLimit() -> UInt32? {
        openForR(CPUFreqFile.biosLimit)
        return files[CPUFreqFile.biosLimit.rawValue]?.readU32()
    }

    /// Get the current frequency.
    public func getCpuinfoCurFreq() -> UInt32? {
        openForR(CPUFreqFile.cpuinfoCurFreq)
        return files[CPUFreqFile.cpuinfoCurFreq.rawValue]?.readU32()
    }

    /// Get the allowable maximum frequency.
    public func getCpuinfoMaxFreq() -> UInt32? {
        openForR(CPUFreqFile.cpuinfoMaxFreq)
        return files[CPUFreqFile.cpuinfoMaxFreq.rawValue]?.readU32()
    }

    /// Get the allowable minimum frequency.
    public func getCpuinfoMinFreq() -> UInt32? {
        openForR(CPUFreqFile.cpuinfoMinFreq)
        return files[CPUFreqFile.cpuinfoMinFreq.rawValue]?.readU32()
    }

    /// Get the transition latency.
    public func getCpuinfoTransitionLatency() -> UInt32? {
        openForR(CPUFreqFile.cpuinfoTransitionLatency)
        return files[CPUFreqFile.cpuinfoTransitionLatency.rawValue]?.readU32()
    }

    /// Get related CPUs.
    public func getRelatedCpus() -> [UInt32]? {
        openForR(CPUFreqFile.relatedCpus)
        return files[CPUFreqFile.relatedCpus.rawValue]?.readU32Arr()
    }

    /// Get the available scaling frequencies.
    public func getScalingAvailableFrequencies() -> [UInt32]? {
        openForR(CPUFreqFile.scalingAvailableFrequencies)
        return files[CPUFreqFile.scalingAvailableFrequencies.rawValue]?.readU32Arr()
    }

    /// Get the available scaling governors.
    public func getScalingAvailableGovernors() -> [String]? {
        openForR(CPUFreqFile.scalingAvailableGovernors)
        return files[CPUFreqFile.scalingAvailableGovernors.rawValue]?.readStringArr()
    }

    /// Get the current scaling frequency.
    public func getScalingCurFreq() -> UInt32? {
        openForR(CPUFreqFile.scalingCurFreq)
        return files[CPUFreqFile.scalingCurFreq.rawValue]?.readU32()
    }

    /// Get the scaling driver.
    public func getScalingDriver() -> String? {
        openForR(CPUFreqFile.scalingDriver)
        return files[CPUFreqFile.scalingDriver.rawValue]?.readString()
    }

    /// Get the scaling governor.
    public func getScalingGovernor(readOnly: Bool = false) -> String? {
        openForRW(CPUFreqFile.scalingGovernor, readOnly)
        return files[CPUFreqFile.scalingGovernor.rawValue]?.readString()
    }

    /// Set the scaling governor.
    public func setScalingGovernor(governor: String, readOnly: Bool = false) -> Bool {
        openForRW(CPUFreqFile.scalingGovernor, readOnly)
        if let f = files[CPUFreqFile.scalingGovernor.rawValue] {
            return f.writeString(governor)
        }
        return false
    }

    /// Get the maximum frequency.
    public func getScalingMaxFreq(readOnly: Bool = false) -> UInt32? {
        openForRW(CPUFreqFile.scalingMaxFreq, readOnly)
        return files[CPUFreqFile.scalingMaxFreq.rawValue]?.readU32()
    }

    /// Set the maximum frequency.
    public func setScalingMaxFreq(freq: UInt32, readOnly: Bool = false) -> Bool {
        openForRW(CPUFreqFile.scalingMaxFreq, readOnly)
        if let f = files[CPUFreqFile.scalingMaxFreq.rawValue] {
            return f.writeU32(freq)
        }
        return false
    }

    /// Get the minimum frequency.
    public func getScalingMinFreq(readOnly: Bool = false) -> UInt32? {
        openForRW(CPUFreqFile.scalingMinFreq, readOnly)
        return files[CPUFreqFile.scalingMinFreq.rawValue]?.readU32()
    }

    /// Set the minimum frequency.
    public func setScalingMinFreq(freq: UInt32, readOnly: Bool = false) -> Bool {
        openForRW(CPUFreqFile.scalingMinFreq, readOnly)
        if let f = files[CPUFreqFile.scalingMinFreq.rawValue] {
            return f.writeU32(freq)
        }
        return false
    }

    /// Set the scaling speed when using the userspace governor.
    public func setScalingSetspeed(freq: UInt32) -> Bool {
        openForW(CPUFreqFile.scalingSetspeed)
        if let f = files[CPUFreqFile.scalingSetspeed.rawValue] {
            return f.writeU32(freq)
        }
        return false
    }

}


Sources/CPUFreqFile.swift
/**
 * CPUFreqFile enumeration
 *
 * @author Connor Imes
 * @date 2017-06-15
 */

/// Enumeration of the files possibly available for each CPU.
public enum CPUFreqFile: Int {
    case affectedCpus = 0
    case biosLimit
    case cpuinfoCurFreq
    case cpuinfoMaxFreq
    case cpuinfoMinFreq
    case cpuinfoTransitionLatency
    case relatedCpus
    case scalingAvailableFrequencies
    case scalingAvailableGovernors
    case scalingCurFreq
    case scalingDriver
    case scalingGovernor
    case scalingMaxFreq
    case scalingMinFreq
    case scalingSetspeed

    /// How many entries are in the enum.
    public static let count: Int = {
        var n: Int = 0
        while let _ = CPUFreqFile(rawValue: n) {
            n += 1
        }
        return n
    }()

    // TODO: Can we roll this into the enum declaration above so we don't need a separate array?
    fileprivate static let fileNames: [String] = [
        "affected_cpus",
        "bios_limit",
        "cpuinfo_cur_freq",
        "cpuinfo_max_freq",
        "cpuinfo_min_freq",
        "cpuinfo_transition_latency",
        "related_cpus",
        "scaling_available_frequencies",
        "scaling_available_governors",
        "scaling_cur_freq",
        "scaling_driver",
        "scaling_governor",
        "scaling_max_freq",
        "scaling_min_freq",
        "scaling_setspeed"
    ]

    internal init?(fileName: String) {
        if let idx = CPUFreqFile.fileNames.index(of: fileName) {
            self = CPUFreqFile(rawValue: idx)!
        } else {
            return nil
        }
    }

    /// Get the file name for a file type.
    public func getFileName() -> String {
        return CPUFreqFile.fileNames[self.rawValue]
    }

    /// Get the complete file path for a file type.
    public func getPath(cpu: UInt32) -> String {
        return "/sys/devices/system/cpu/cpu\(cpu)/cpufreq/\(getFileName())"
    }
}
